content/blog/trungnt2910/2024-06-28_[gsoc_2024]_debugging_progress_2.md

```diff
@@ -0,0 +1,485 @@
+
+Many GDB use cases involve setting breakpoints on the main image before execution. Without knowing
+the offsets, GDB would try to insert breakpoints at addresses relative to `0x0`. When the target
+starts execution, the debugger would try to write to these invalid addresses. It would then fail,
+preventing execution from continuing.
+
+To prevent this issue, for targets with relocatable main executables, GDB provides
+`disable_breakpoints_before_startup` to delay committing breakpoints into the target address
+space. When relocation is finished, the handler can call `enable_breakpoints_after_startup`.
+
+Haiku's `nat` code calls `disable_breakpoints_before_startup` right after a new inferior is created
+and when the main executable is unloaded (which usually means an upcoming `exec` event). Then, after
+the debugger successfully attaches or handles the `exec` event, the `nat` code will re-enable the
+breakpoints. This will only be done if the current executable reported by the target's OS matches
+the one known by GDB. Otherwise, if the inferior if executed through a wrapper shell, GDB might
+commit the breakpoints too early, corrupting the address space of that shell.
+
+## "Non-stop" and "All-stop" debugging
+
+By default, GDB functions in "all-stop" mode. When a thread hits a debugger event, the whole process
+is stopped. When the user wishes to resume, the whole process is resumed.
+
+["Non-stop"](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Non_002dStop-Mode.html#Non_002dStop-Mode)
+mode does not stop the whole process when a thread hits an event. All non-signalled threads continue
+running as usual. Because the other threads are still running, many operations such as inspecting
+registers or the stack trace would not work.
+
+Haiku's Debugger API only works in non-stop mode. However, debugging in non-stop mode or emulation
+of all-stop on top of non-stop in GDB is tied to another feature, asynchronous event handling. This
+feature requires all debugger events to be representable as a `poll`able file descriptor. The `wait`
+`target_op` method will not be called unless `poll` marks the provided FD.
+
+On `ptrace`-based OSes, this is done by handling the `SIGCHLD` sent to the tracer every time a
+tracee hits an event. The signal handler writes a pipe, marking the FD as ready for the next `poll`
+in the event loop.
+
+Haiku does not send the debugger `SIGCHLD` or similar signals. Furthermore, ports are used instead
+of FDs for communicating with the debugger. A potential solution is to put all open debugger ports
+into an `event_queue`, then return the file descriptor of that queue to the event loop. However,
+this file descriptor type on Haiku currently
+[does not](https://github.com/haiku/haiku/blob/2d24dd21846e3d71961c5d84190d235397c0d9ae/src/system/kernel/events/event_queue.cpp#L599-L600)
+support polling (please upvote [this ticket](https://dev.haiku-os.org/ticket/18954) if you are
```

Comment:
This is a pretty technical detail and turned out to not be necessary to port GDB's non-stop mode. I don't think it makes sense to ask users to upvote the ticket. Linking it and mentioning the discussion may make sense, at most.

Reply:
Removed the call to upvote that ticket, but:

> turned out to not be necessary

Still necessary to cleanly port it. The current solution is more like a temporary hack.

Spawning threads in otherwise single-threaded programs have implications, especially in a program that uses `fork` here and there. Normally I would avoid threads as much as possible and use an event-based asynchronous approach (either signals or polling).

> This is a pretty technical detail

The whole blog is meant to be a lot of technical details...